### DISCUSSION TITLE
Алгоритм поиска треков для кэша истории Lastfm

### DISCUSSION THREAD

**Discussion by efremych** [Chimildic/goofy#30] [Помощь с ошибкой] [answered]
**Проигранный трек (из кэша Spotify)**
{"track":{"album":{"album_type":"album","artists":[{"id":"4whxX965DKKqYkXmk33b4E","name":"Guano Apes"}],"id":"1N5L209IqXQ6cqV2HFJU4G","name":"Proud Like a God","release_date":"1997","release_date_precision":"year","total_tracks":12},"artists":[{"id":"4whxX965DKKqYkXmk33b4E",**"name":"Guano Apes"**},{"id":"318EDeSl1oHUaOGbIvXB3t",**"name":"Ronald Prent at Wisselord Studios, Hilversum, Netherland**"}],"duration_ms":202093,"explicit":false,"id":"5yhYJs4S7hucEeeLbyW0MA",**"name":"Get Busy"**,"popularity":23}

**Проигранный трек согласно кэшу Last FM**
{"album":{"album_type":"album","artists":[{"id":"4whxX965DKKqYkXmk33b4E","name":"Guano Apes"}],"id":"3d5uCI9aMb09ECHj37ccZr","name":"Proud Like a God XX","release_date":"2017-10-06","release_date_precision":"day","total_tracks":20},"artists":[{"id":"4whxX965DKKqYkXmk33b4E",**"name":"Guano Apes**"}],"duration_ms":206906,"explicit":false,"id":"0icMydCjioOzYW9O0ibiOX",**"name":"Get Busy - 2017 Mix"**,"popularity":23,"played_at":"26 Jan 2021, 10:24","source":"Last FM Station undefined"},

Почему для кэша Last FM был выбран трек, который в результатах поиска располагается ниже реально проигранного?
![image](https://user-images.githubusercontent.com/76212904/105881848-0c7d0900-6016-11eb-80eb-c45b0f16d3b5.png)

**Comment by efremych** [comment, marked as answer]
Вопрос снимается)
Блин, так Last FM отдает (почему-то)...

 disc_number: 1,
  duration_ms: 206906,
  explicit: false,
  external_ids: { isrc: 'DEA811700724' },
  external_urls: { spotify: 'https://open.spotify.com/track/0icMydCjioOzYW9O0ibiOX' },
  href: 'https://api.spotify.com/v1/tracks/0icMydCjioOzYW9O0ibiOX',
  id: '0icMydCjioOzYW9O0ibiOX',
  is_local: false,
  name: 'Get Busy - 2017 Mix',
  popularity: 23,
  preview_url: 'https://p.scdn.co/mp3-preview/a84a615d62dd4da36a9f804fbf74ea2002be9c42?cid=a705dcf6a1204a1fa66ec9eff942c0b1',
  track_number: 9,
  type: 'track',
  uri: 'spotify:track:0icMydCjioOzYW9O0ibiOX',
  played_at: '26 Jan 2021, 10:24',
  source: 'Last FM Station recent' }


Если быть точным, виной всему сам спотифай :( Выдает в результатах поиска то, что считает нужным...
![image](https://user-images.githubusercontent.com/76212904/105907625-3fcf9000-6036-11eb-99b8-fb1af47d26fb.png)
![image](https://user-images.githubusercontent.com/76212904/105907720-5544ba00-6036-11eb-8060-1a473176f2ac.png)

**Comment by efremych** [comment]
А есть ли какая-то возможность избежать такого?

Поисковый запрос трека `mutter rammstein mutter` отдает результат  `"name": "Sonne"`. При этом в десктопе и мобильном приложении первым корректно выдается трек Mutter.

Я добавил в запрос помимо артиста трека, еще и альбом (альбом-артист-трек), так как комбинация артист-трек выдавала не всегда нужный результат. А, получается, у любой комбинации свои минусы...

При этом, в таком виде Консоль выдает корректный результат `album:mutter artist:rammstein track:mutter `.
И так тоже корректно `track: mutter rammstein mutter`.

Правда, некоторые другие треки тогда не находятся...
Или идеального варианта нет? :( И текущий вариант - золотая середина

**Reply by Chimildic** [reply]
Избежать где? Для чего вам такой точеный поиск по заранее известному треку?

**Reply by efremych** [reply]
Для истории прослушивания Last Fm. Как показала практика, лучше ориентироваться на неё - История Spotify сбоит чаще и восстановить ее проблематичнее из-за ограничения в 50 треков. 

Я теперь правда не уверен, что и трюк с попыткой добавления названия альбома в поисковый запрос даёт нужный результат, раз в Приложении и в Консоли разные результаты :(
А добавлял я Альбом потому что в Историю попадал трек с совершенно другой длительностью с другого альбома.

Трек выше стал известен, когда я начал разбираться почему левый трек попадает в Историю. До этого он, можно сказать, неизвестен был)

Конечно, и у Last Fm свои недостатки есть. Встречал треки, которые наверно определялись (видимо из-за недочётов в базе MusicBrainz, хз). 

У Spotify встречал орфографические ошибки в названиях треков. Также, заметил что замена символов на пробелы может испортить поиск названия трека, при том что на поиск по исполнителю это не влияет.

Но вот задался вопросом, возможно ли хотя бы как-то уточнить поиск по данным от Last Fm…

**Reply by Chimildic** [reply]
Можно на свежих примерах? Я не понимаю о чем конкретно речь. 

Если last.fm дает `X mix`, а был прослушан просто `X`. Теоретически, можно отбросить `mix` и искать оригинал. Но когда действительно был прослушан `X mix` - его уже не найти. То есть у нас нет источника достоверной информации о прослушанной версии трека. 

В библиотеке предпринята попытка поиска по равенству среди результатов от Spotify `findBest`. Формируется ключи из исполнителя и названия трека, которые сравниваются со значением от last.fm. Когда нужный трек не первый в поиске Spotify и last.fm дал имя без искажений, скорее всего вернется трек, который был действительно прослушан. Перед сравнением имена очищаются `formatName`. Но при опечатках и разнице написании через спец. символы способ не сработает. Потому что в одном случае спец. символы заменятся, а в другом случае останется обычная буква.

О каком проценте ошибок поиска вообще идет речь?

Не уверен, что теги типа `album:`, `track:` и `artist:` имеют смысл. Потому что они не отражены в справке Spotify.

**Reply by efremych** [reply]
> О каком проценте ошибок поиска вообще идет речь?

Это единичные случаи, да и сложно их заметить, это надо следить за тем что слушаешь, всего ведь не упомнишь. Что попалось на глаза.

> Можно на свежих примерах? Я не понимаю о чем конкретно речь.

ну, ради интереса

https://open.spotify.com/track/4HcGbZ4n4j1rObwTmix9Ca?si=f2a49ccb5ec14c0e
если правильно помню, возвращается первый трек из результата, не проигранный этот.
Отдавал бы Last FM длительность трека еще, можно было бы наверно уточнить лучший результат...

https://open.spotify.com/track/5BVNXl23xEtFgsKmbKpYfA?si=f510c8a2962f4101
здесь ошибка в самом last fm (это полагаю никак не победить), плюс извращения над названием трека от авторов и определяется проигранным вот этот трек https://open.spotify.com/track/2qOmcSjOQEDIJKosonn75a?si=7b52c4685d2e4aba

https://open.spotify.com/track/3qppQHKKruwsaiD7EhBKKK?si=08a731d91f6c4a12
этот трек требователен к точкам в названии (с пробелами трек не будет найден, я пока у себя убрал пробелы для точек)

> Поисковый запрос трека mutter rammstein mutter отдает результат "name": "Sonne". При этом в десктопе и мобильном приложении первым корректно выдается трек Mutter.

Ну а здесь у меня был такой поиск. Думал разгадал загадку поиска в Spotify, но как оказалось - нет
`return `${formatAlbumNameLastfm(item)} ${item.name}`.formatName();
`
> Не уверен, что теги типа album:, track: и artist: имеют смысл. Потому что они не отражены в справке Spotify.

Пока что оставил у себя теги, до очередной ошибки поиска))) Каждый раз кажется - вот оно. нашел шаблон, ан нет...
Не все треки правда так ищутся (странный поиск у Spotify). Но для второго захода оставил поиск по стандартному шаблону. В 2 подхода, так сказать...

Перфекционизм, будь он не ладен...

**Reply by Chimildic** [reply]
По идеи `Hypa Hypa` должно отфильтроваться по названию при `removeTracks`. А если вы зачем-то ведете плейлист истории, то вряд ли найдется должный способ.

Какой клиент скроббла отслеживает Spotify? Если нативный из настроек lastfm, попробуйте сторонний настроить. На предмет того, чтобы не было автокоррекции названий со стороны lastfm. 

Альтернативный вариант накопления истории есть в [теме](https://github.com/Chimildic/goofy/discussions/53). Но там свои сложности и недостатки.

**Reply by efremych** [reply]
> По идеи Hypa Hypa должно отфильтроваться по названию при removeTracks

Там длительности треков разные. 
Пока писал, вспомнил, что это я вроде длительность трека у себя добавил для removeTracks. Ну не убирать же теперь) иначе прослушал одну Hypa Hypa, а удалились обе из какого нибудь плейлиста. Плюс на длительности трека у меня завязан огород Истории прослушивания (не буду вдаваться в подробности :)) 
Или например топ прослушиваний, длительность помогает различить треки при подсчёте - так вот Hypa Hypa заметил. 

> А если вы зачем-то ведете плейлист истории, то вряд ли найдется должный способ

Плейлист лишь зеркало. История от Spotify всем хороша и в идеале иметь ее, но туда попадают треки с паузы и она крайне не стабильная. 

> попробуйте сторонний настроить

да как-то привык уже… Подумаю конечно, может и правда что есть…

> Альтернативный вариант накопления истории

да, что-то слышал, но я так понял это или для ПК или для андроида. К сожалению, одно не нужно, другого нету…

По сути, тут вопросы к немного загадочному алгоритму поиска Spotify и данным от last fm (могли бы и побольше данных давать). Но скорее риторические вопросы.

**Reply by Chimildic** [reply]
> Там длительности треков разные.

Стандартный `removeTracks` не смотрит на длительность. Она лишь в поиске дубликатов учитывается.

> или для ПК или для андроида

Можно настроить и то, и другое одновременно. Главный фокус в том, как автоматически доставлять файл до Google Диска с этих устройств. Если в наличии только iOS, то не представляю возможно ли такое. Может в нем есть доступ к файлам приложений и программы типа Tasker'а.

> загадочному алгоритму поиска Spotify

Наверное более первична проблема достоверности данных от last.fm. Если скроблер будет игнорировать автокоррекцию и создавать, пусть дубликаты для базы last.fm, но точные названия от Spotify - ситуация должна стать лучше. 

И касаемо тысячи версий одного трека - Spotify то решить проблему не может, у нас тем более не выйдет.
